src/core/dbus-execute.c

```diff
@@ -1423,6 +1423,10 @@ int bus_set_transient_exec_command(
                 if (r < 0)
                         return r;
 
+                if (!argv)
+                        return sd_bus_error_setf(error, SD_BUS_ERROR_INVALID_ARGS,
+                                                 "\"%s\" argv cannot be empty", name);
+
```

Comment:
I think we should accept an empty argv here. The empty argv means resetting exec command.

Reply:
Are you sure ? The signature is typically `a(sasb)`:
```
ExecStartPre, ExecStart, ExecStartPost, ExecReload, ExecStop, and ExecStop are arrays of structures where each struct contains: the binary path to execute; an array with all arguments to pass to the executed command, starting with argument 0; a boolean whether it should be considered a failure if the process exits uncleanly; two pairs of CLOCK_REALTIME/CLOCK_MONOTONIC microsecond timestamps when the process began and finished running the last time, or 0 if it never ran or never finished running; the PID of the process, or 0 if it has not run yet; the exit code and status of the last run. This field hence maps more or less to the corresponding setting in the service unit file but is augmented with runtime data.
```

Did you mean that the `a(sasb)` array itself can be empty ? this is looking at the inner `as` array.

Reply:
Also of course this change does fix the crash described in https://github.com/systemd/systemd/issues/20933

```
busctl call \
  org.freedesktop.systemd1 /org/freedesktop/systemd1 \
  org.freedesktop.systemd1.Manager StartTransientUnit \
  "ssa(sv)a(sa(sv))" test.service replace \
  1 ExecStart "a(sasb)" 1 /usr/bin/sleep 0 true 0
```
now fails with a dbus error instead of crashing, and e.g.
```
busctl call \
  org.freedesktop.systemd1 /org/freedesktop/systemd1 \
  org.freedesktop.systemd1.Manager StartTransientUnit \
  "ssa(sv)a(sa(sv))" test.service replace \
  1 ExecStart "a(sasb)" 1 /usr/bin/sleep 2 /usr/bin/sleep infinity true 0
```
still works

Reply:
Ah, the dbus method accepts `path` and `argv` separately. Hm, then this may be good. Sorry for the noise.